File: modules/device-id/src/DeviceIdModule.ts
import { requireNativeModule } from 'expo-modules-core';

// This name "DeviceId" MUST match the Name("DeviceId") in your Kotlin file
const DeviceId = requireNativeModule('DeviceId');

export function getAndroidId(): string {
  // Check if DeviceId exists to prevent the 'undefined' crash
  if (!DeviceId || !DeviceId.getAndroidId) {
    console.warn("⚠️ DeviceId native module is not linked yet.");
    return "unknown";
  }
  return DeviceId.getAndroidId();
}

File: modules/device-id/android/src/main/java/expo/modules/deviceid/DeviceIdModule.kt
package expo.modules.deviceid

import android.provider.Settings
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class DeviceIdModule : Module() {
  override fun definition() = ModuleDefinition {
    // 1. This NAME must match what you call in requireNativeModule
    Name("DeviceId")

    // 2. Use Function for synchronous calls
    Function("getAndroidId") {
      // Get the content resolver from the current activity/context
      val contentResolver = appContext.reactContext?.contentResolver
      return@Function Settings.Secure.getString(
        contentResolver,
        Settings.Secure.ANDROID_ID
      ) ?: "unknown"
    }
  }
}